In [56]:
import pickle
import pandas as pd


In [57]:
#opening our clustered model
with open('finalized_model.sav', 'rb') as fid:
    model = pickle.load(fid)


In [59]:
data = pd.read_csv("Opportunity 2.csv")


In [60]:
data = data.dropna(how='any')


In [61]:
#Converting the whole dataset into smaller case
data=data.apply(lambda x: x.astype(str).str.lower())


In [62]:
comb_frame = data['Discipline'].str.cat(" "+data['Sub tags'])


In [63]:
from sklearn.cluster import KMeans
from sklearn import metrics
from sklearn.feature_extraction.text import TfidfVectorizer


In [64]:
vectorizer = TfidfVectorizer(stop_words='english')
X = vectorizer.fit_transform(comb_frame)


In [65]:
label_dict = {
                 0: "Social Humanities",
                 1: "Management",                  
                 2: "Engineering & Technology  science",
                
            }


In [67]:
# load the complete data in a dataframe
data=pd.read_csv("Opportunity 2.csv")

#Converting the whole dataset into smaller case
data=data.apply(lambda x: x.astype(str).str.lower())

data=data.dropna(how='any')

# create new column in dataframe which is combination of (ID, Sub tags, Discipline) in existing data-frame
data['InputString'] = data['Discipline'].str.cat(" "+data['Sub tags'])

data['ClusterPrediction'] = ""


def cluster_predict(str_input):
    Y = vectorizer.fit_transform(list(str_input))
    prediction = model.predict(Y)
    return prediction

# Cluster category for each live course
data['ClusterPrediction']=data.apply(lambda x: cluster_predict(data['InputString']), axis=0)


In [69]:

# load the complete data in a dataframe
data=pd.read_csv("Opportunity 2.csv")

#Converting the whole dataset into smaller case
data=data.apply(lambda x: x.astype(str).str.lower())

data=data.dropna(how='any')

# create new column in dataframe which is combination of (ID, Sub tags, Discipline) in existing data-frame
data['InputString'] = data['ID'].str.cat(" "+data['Sub tags'].str.cat(" "+data['Discipline']))

data['ClusterPrediction'] = ""


def cluster_predict(str_input):
    Y = vectorizer.transform(list(str_input))
    prediction = model.predict(Y)
    return prediction

# Cluster category for each live course
data['ClusterPrediction']=data.apply(lambda x: cluster_predict(data['InputString']), axis=0)


In [70]:
def recommend_util(str_input):
    
    # match on the basis ID 
    temp_df = data.loc[data['ID'] == str_input]
    temp_df['InputString'] = temp_df['ID'].str.cat(" "+temp_df['Sub tags'].str.cat(" "+temp_df['Discipline']))
    str_input = list(temp_df['InputString'])
    
    
    # Predict category of input string category
    prediction_inp = cluster_predict(str_input)
    prediction_inp = int(prediction_inp)
    # Based on the above prediction 10 random courses are recommended from the whole data-frame
    # Recommendation Logic is kept super-simple for current implementation.
    temp_df = data.loc[data['ClusterPrediction'] == prediction_inp]
    temp_df = temp_df.sample(2)
    
    return list(temp_df['ID'])



In [71]:
if __name__ == '__main__':
    queries = ['1','50']
for query in queries:
    res = recommend_util(query)
    print(res)


['267', '1471']
['127', '1884']


C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


For <b>Query 1-'1'</b>: It recommended us ['267', '1471'] this array.


<br>
<b>1</b>	-International Fellowship World Forestry Center
<br><b>Deadline:</b>	10/15/2019	<b>Link:</b>https://www.noticebard.com/call-for-applications-international-fellowship-world-forestry-center/  <b>Type of Opportunity:</b>Conferences		<b>Discipline:</b>Humanities and Social Sciences	<b>Minimum Qualification / Applicable for:</b>Bachelors	<b>Countries Eligible:</b>ALL	<b>Location of opportunity:</b>USA
		

    
<br>
<br>
<i>The recommended options are:</i>

<b>267</b>	-Getty/ACLS Postdoctoral Fellowships in the History of Art		<br><b>Funded:</b>Fully funded	<b>Deadline:</b>10/23/2019	<b>Link:</b>https://ofa.acls.org/	<b>Type of Opportunity:</b>Fellowships, Scholarships		<b>Discipline:</b>Humanities and Social Sciences	<b>Minimum Qualification / Applicable for:</b>Doctorate	   <b>Countries Eligible:</b>All <b>Location of opportunity:</b>	USA

<br>
<br>
<b>1471</b>	-Humane Studies Fellowship 2020-2021		<br>	<b>Deadline:</b>12/1/2019	<b>Link:</b>https://opportunitydesk.org/2019/11/18/humane-studies-fellowship-2020-2021/	<b>Type of Opportunity:</b>Fellowships		<b>Discipline:</b>Humanities and Social Sciences	<b>Minimum Qualification / Applicable for:</b>Post Doctorate	   <b>Countries Eligible:</b>All <b>Location of opportunity:</b>	USA
<br>




For <b>Query 2-'50'</b>: It recommended us ['127', '1884'] this array.


<br>
<b>50</b>	-Kectil Leadership Program 2020 for Young Leaders in Developing Countries
<br><b>Funded:</b>Partially Funded	 <b>Deadline:</b>	10/31/2019	<b>Link:</b>https://kectil.com/holding-page-123/  <b>Type of Opportunity:</b>Fellowships			<b>Discipline:</b>All	<b>Minimum Qualification / Applicable for:</b>All	<b>Countries Eligible:</b>ALL	<b>Location of opportunity:</b>Online
		

<br>
<br>
<i>The recommended options are:</i>

<b>127	</b>	-Aspen Institute’s New Voices Fellowship 2020 for Development Professionals	<br><b>Funded:</b>Fully funded	<b>Deadline:</b>10/15/2019<b>Link:</b>http://newvoicesfellows.aspeninstitute.org/Nominations<b>Type of Opportunity:</b>Fellowships			<b>Discipline:</b>All	<b>Minimum Qualification / Applicable for:</b>All	   <b>Countries Eligible:</b>All <b>Location of opportunity:</b>	USA


<br>
<br>
<b>1884</b>	-Amgen Scholars Asia Program 2020		<br>	<b>Deadline:</b>	2/3/2020	<b>Link:</b>http://amgenscholars.com/asia-program/	<b>Type of Opportunity:</b>Fellowships		<b>Discipline:</b>All	<b>Minimum Qualification / Applicable for:</b>	Bachelors	   <b>Countries Eligible:</b>All <b>Location of opportunity:</b>	None
<br>

	
